docs/source/design/result_records.rst

```diff
@@ -0,0 +1,209 @@
+Each command invocation yields one or more result records. Result records are
+routinely inspected throughout the code base, and are used to inform generic
+error handling, as well as particular calling commands on how to proceed with
+a specific operation.
+
+The technical implementation of a result record is a Python dictionary.  This
+dictionary must contain a number of mandatory fields/keys (see below). However,
+an arbitrary number of additional fields may be added to a result record.
+
+The ``get_status_dict()`` function simplifies the creation of result records.
+
+.. note::
+   Developers *must* compose result records with care! DataLad supports custom
+   user-provided hook configurations that use result record fields to
+   decide when to trigger a custom post-result operation. Such custom hooks
+   rely on a persistent naming and composition of result record fields.
+   Changes to result records, including field name changes, field value changes,
+   but also timing/order of record emitting potentially break user set ups!
+
+
+Mandatory fields
+================
+
+The following keys *must* be present in any result record. If any of these
+keys is missing, DataLad's behavior is undefined.
+
+
+``action``
+----------
+
+A string label identifying which type of operation a result is associated with.
+Labels *must not* contain white space. They should be compact, and lower-cases,
+and use ``_`` (underscore) to separate words in compound labels.
+
+A result without an ``action`` label will not be processed and is discarded.
+
+
+``path``
+--------
+
+A string with an *absolute* path describing the local entity a result is
+associated with. Paths must be platform-specific (e.g., Windows paths on
+Windows, and POSIX paths on other operating systems). When a result is about an
+entity that has no meaningful relation to the local file system (e.g., a URL to
+be downloaded), to ``path`` value should be determined with respect to the
+potential impact of the result on any local entity (e.g., a URL downloaded
+to a local file path, a local dataset modified based on remote information).
+
+
+``status``
+----------
+
+This field indicates the nature of a result in terms of four categories, identified
+by a string label.
+
+- ``ok``: a standard, to-be-expected result
+- ``notneeded``: an operation that was requested, but found to be unnecessary
+  in order to achieve a desired goal
+- ``impossible``: a requested operation cannot be performed, possibly because
+  its preconditions are not met
+- ``error``: an error occurred while performing an operation
+
+Based on the ``status`` field, a result is categorized into *success* (``ok``,
+``notneeded``) and *failure* (``impossible``, ``error``). Depending on the
+``on_failure`` parameterization of a command call, any failure-result emitted
+by a command can lead to an ``IncompleteResultsError`` being raised on
+command exit, or a non-zero exit code on the command line.
+
```

Comment:
In addition to raising errors, isn't it also "nothing" when we set ``on_failure='ignore'``? Could we maybe add the possible values that `on_failure` can take (sorry if I have missed them)?

Reply:
Will do, good point, thx!